File: Data Structures & Algorithms/three-integer-sum/submission-2.py
class Solution:
    def threeSum(self, nums: List[int]) -> List[List[int]]:
        nums.sort()
        ans = list()

        print(nums)

        for i in range(len(nums) - 2):
            print(nums[i])
            if i > 0:
                if nums[i-1] == nums[i]:
                    continue
            
            l = i+1
            r = len(nums) -1
            while l < r:
                if nums[l] + nums[r] == -1 * nums[i] :
                    ans.append([nums[i], nums[l], nums[r]])
                    l +=1
                    while nums[l] == nums[l-1] and l < r:
                        l += 1
                elif nums[l] + nums[r] < -1 * nums[i]:
                    l+=1
                else:
                    r -= 1
        return ans
        
        

File: Data Structures & Algorithms/partition-equal-subset-sum/submission-2.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
       int sum =0;
       for(const auto& x : nums)
       {
            sum +=x;
       } 
       if(sum%2 == 1)return false;
       int sum1 = sum/2;
       int sum2 = sum/2;
       sort(nums.begin(), nums.end());
       for(int i=nums.size()-1; i >=0; i--)
       {
            if(sum1 >= nums[i]) sum1-=nums[i];
            else if(sum2 >= nums[i]) sum2-=nums[i];
            else return false;
       }
       if(sum1==0 && sum2 == 0)return true;
       else return false;
    }
};
